File: ming/0927/LowestCommonAncestorofaBinarySearchTree.java

public class LowestCommonAncestorofaBinarySearchTree {
	
	
	public TreeNode lowestCommonAncestorBST(TreeNode root, TreeNode p, TreeNode q) {
		if (root==null || p == null || q == null) return null;
		if (root.val == p.val || root.val == q.val) return root;
		if (root.val > p.val && root.val > q.val)
			return lowestCommonAncestorBST(root.left, p, q);
		if (root.val < p.val && root.val< q.val)
			return lowestCommonAncestorBST(root.right, p, q);
		return root;		
    }
	
	public TreeNode lowestCommonAncestorBT(TreeNode root, TreeNode p, TreeNode q) {
		if (root==null || p == null || q == null) return null;
		if (root == p || root == q) return root;
		
		TreeNode lRes = lowestCommonAncestorBT(root.left, p, q);
		TreeNode rRes = lowestCommonAncestorBT(root.right, p, q);
		
		if (lRes == null) return rRes;
		if (rRes == null) return lRes;
		
		return root;
    }

}

File: Sen/Lowest_Common_Ancestor_of_a_Binary_Search_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @param {TreeNode} p
    # @param {TreeNode} q
    # @return {TreeNode}
    def lowestCommonAncestor(self, root, p, q): # iteration
    	while root:
            if max(p.val, q.val) < root.val:
                root = root.left
            elif min(p.val, q.val) > root.val:
                root = root.right
            else:
                return root
        return None

    def lowestCommonAncestor(self, root, p, q): # recursion
        if not root:
            return None
        if root.val > max(p.val, q.val):
            return self.lowestCommonAncestor(root.left, p, q)
        elif root.val < min(p.val, q.val):
            return self.lowestCommonAncestor(root.right, p, q)
        else:
            return root